drivers/charger/bq24190.c

```diff
@@ -0,0 +1,491 @@
+		*online = CHARGER_ONLINE_OFFLINE;
+	}
+
+	return 0;
+}
+
+static int bq24190_charger_get_status(const struct device *dev, enum charger_status *status)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint8_t ss_reg, chrg_fault;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_F, &chrg_fault);
+	if (ret) {
+		return ret;
+	}
+
+	chrg_fault = FIELD_GET(BQ24190_REG_F_CHRG_FAULT_MASK, chrg_fault);
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_SS, &ss_reg);
+	if (ret) {
+		return ret;
+	}
+
+	/*
+	 * The battery must be discharging when any of these are true:
+	 * - there is no good power source;
+	 * - there is a charge fault.
+	 * Could also be discharging when in "supplement mode" but
+	 * there is no way to tell when its in that mode.
+	 */
+	if (!(ss_reg & BQ24190_REG_SS_PG_STAT_MASK) || chrg_fault) {
+		*status = CHARGER_STATUS_DISCHARGING;
+	} else {
+		ss_reg = FIELD_GET(BQ24190_REG_SS_CHRG_STAT_MASK, ss_reg);
+
+		switch (ss_reg) {
+		case BQ24190_CHRG_STAT_NOT_CHRGING:
+			*status = CHARGER_STATUS_NOT_CHARGING;
+			break;
+		case BQ24190_CHRG_STAT_PRECHRG:
+		case BQ24190_CHRG_STAT_FAST_CHRG:
+			*status = CHARGER_STATUS_CHARGING;
+			break;
+		case BQ24190_CHRG_STAT_CHRG_TERM:
+			*status = CHARGER_STATUS_FULL;
+			break;
+		default:
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+static int bq24190_charger_get_ccc(const struct device *dev, uint32_t *curr)
+{
+	const struct bq24190_config *const config = dev->config;
+	bool frc_20pct;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_CCC, &v);
+	if (ret) {
+		return ret;
+	}
+
+	frc_20pct = v & BQ24190_REG_CCC_FORCE_20PCT_MASK;
+
+	v = FIELD_GET(BQ24190_REG_CCC_ICHG_MASK, v);
+
+	*curr = (v * BQ24190_REG_CCC_ICHG_STEP_UA) + BQ24190_REG_CCC_ICHG_OFFSET_UA;
+
+	if (frc_20pct) {
+		*curr /= 5;
+	}
+
+	return 0;
+}
+
+static int bq24190_charger_get_pc(const struct device *dev, uint32_t *pc)
+{
+	const struct bq24190_config *const config = dev->config;
+	bool frc_20pct;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_CCC, &v);
+	if (ret) {
+		return ret;
+	}
+
+	frc_20pct = v & BQ24190_REG_CCC_FORCE_20PCT_MASK;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_PCTCC, &v);
+	if (ret) {
+		return ret;
+	}
+
+	v = FIELD_GET(BQ24190_REG_PCTCC_IPRECHG_MASK, v);
+
+	*pc = (v * BQ24190_REG_PCTCC_IPRECHG_STEP_UA) + BQ24190_REG_PCTCC_IPRECHG_OFFSET_UA;
+
+	if (frc_20pct) {
+		*pc /= 2;
+	}
+
+	return 0;
+}
+
+static int bq24190_charger_get_tc(const struct device *dev, uint32_t *tc)
+{
+	const struct bq24190_config *const config = dev->config;
+	uint8_t v;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ24190_REG_PCTCC, &v);
+	if (ret) {
+		return ret;
+	}
+
+	v = FIELD_GET(BQ24190_REG_PCTCC_ITERM_MASK, v);
+
+	*tc = (v * BQ24190_REG_PCTCC_ITERM_STEP_UA) + BQ24190_REG_PCTCC_ITERM_OFFSET_UA;
+
+	return 0;
+}
+
+static int bq24190_get_cvc(const struct device *dev, uint32_t *voltage)
```

Comment:
Keep the unit? `uint32_t voltage_uv` or something, same for the others. That way it's easier to review that the units are right.

Reply:
ack